### ISSUE TITLE
Documentation comments inconsistencies / bugs

### ISSUE THREAD

**Issue by gmpreussner** [nim-lang/Nim#1799]
These work:

    type Foo = object
      ## win

    proc foo() =
      ## win

    iterator bar(..) =
      ## win

However, this doesn't compile (Error: identifier expected, but found '## fail'

    type Bar = enum
      ## fail 

Interestingly, this works though:

    type Foobar = enum ## win

**Comment by gmpreussner** [comment]
This is actually broken altogether. The following code...

```nimrod
type Foo = enum
  one, ## one
  two, ## two
  three, ## three
```

... will generate the following documentation:

    _________________________
    Foo = enum
      one, two, ## two
      three     ## three
    _________________________
    one

**Comment by gmpreussner** [comment]
workaround:

```nimrod```
type Foo = enum ## foo
  one, ## one
  two, ## two
  three, ## three
```

**Comment by gmpreussner** [comment]
Template comments don't show up at all:

    proc foo*() =
      ## this works

    template foobar*() =
      ## will not generate documentation

**Comment by gmpreussner** [comment]
For type related comment issues see: https://github.com/Araq/Nim/issues/1929

**Comment by gmpreussner** [comment]
Use between object and tuple is also inconsistent (same as enum):

```Nimrod
Foo = object
  ## this works

Bar = tuple
  ## this does not work
```

workaround is the same as enum:

```Nimrod
Bar = tuple ## \
  ## works
```

**Comment by gmpreussner** [comment]
I wrote a unit test for this Issue, but I'm not sure where to put it. Would this go into /parser or /misc?

```nim
proc MyProc1*() = ## Comment behind procedure
  discard

proc MyProc2*() =
  ## Comment below procedure
  discard


template MyTemplate1*() = discard ## Comment behind template

template MyTemplate2*() = discard
  ## Comment below template


const
  MyConst1* = 1 ## Comment behind constant
  MyConst2* = 2
    ## Comment below constant


var
  MyVar1* = 1 ## Comment behind variable
  MyVar2* = 2
    ## Comment below variable


type
  MyObject1* = object
    ## Comment below declaration
    field1*: int ## Comment behind field
    field2*: int ## Comment behind field
    field3*: int
      ## Comment below field
    field4*: int
      ## Comment below field

  MyObject2* = object ## Comment behind declaration
    field1*: int


type
  MyTuple1* = tuple
    ## Comment below declaration
    field1: int ## Comment behind field
    field2: int ## Comment behind field
    field3: int
      ## Comment below field
    field4: int
      ## Comment below field

  MyTuple2* = tuple ## Comment behind declaration
    field1: int


type
  MyEnum1* = enum
    ## Comment below declaration
    value1, ## Comment behind value
    value2,
      ## Comment below value with comma
    value3
      ## Comment below value without comma

  MyEnum2 = enum ## Comment behind declaration
    value1
```

The following problems are still present:

- comment below enum, tuple does not compile
- comment below object field only compiles for first field that has a below-comment
- comment below tuple field does not compile for any fields
- comment below enum value only works if trailing comma is present
- template comment doesn't generate documentation

Workarounds exist via the "## \" trick, but I feel that this should really be fixed for 1.0 as it makes the syntax look somewhat unprofessional. Writing comments is one of the first things a new user will do, and if that already doesn't compile, then how much trust can they put into the rest of the language?

**Comment by dom96** [comment]
yeah, this should definitely be fixed. Thank you for investigating. I'm afraid I don't know where you should put the test, @Araq?

**Comment by Araq** [comment]
Test under parser/, though there a couple of bugs/glitches listed here, not everything is a parser problem.

**Comment by Araq** [comment]
We should really fix this for 0.12.0.
